cpp/src/arrow/compute/exec.cc

```diff
@@ -498,15 +510,36 @@ struct NullGeneralization {
     return PERHAPS_NULL;
   }
 
+  static type Get(const ChunkedArray& chunk_array) {
+    if (chunk_array.num_chunks() == 0 ||
```

Comment:
if `num_chunks() == 0`, we should return `ALL_VALID` instead of `ALL_NULL` or `PERHAPS_NULL` like `Get` does above (it doesn't check `.length()` at all which I think was a missed opportunity of making `ALL_VALID` more likely).